### PR TITLE
Fix: Fixed a bug in regards to empty inputs in AddTextLetterByLetter class. 

### PR DESCRIPTION
**Pull request by Immanuel-Alvaro-Bhirawa**
## Overview: What does this pull request change?

Overview
This PR addresses a specific edge case error that occurs when the AddTextLetterByLetter class is used with an empty input. Previously, using an empty input would result in an "Output file does not contain any stream" error, caused by the class attempting to render a video with a run_time of 0.

Changes Made
I have added an exception catch for this edge case to provide better error feedback. If the input is empty, a ValueError is raised with a descriptive message, preventing the error from occurring and informing the user of the issue

## Motivation and Explanation: Why and how do your changes improve the library?
 The motivation behind this change is to improve the user experience by offering clearer error feedback and preventing an unintuitive error message. This will aid Manim users and developers who have encountered this issue, as reported in git issues.

## Links to added or changed documentation pages
 PR Link: https://github.com/ManimCommunity/manim/pull/3404

## Further Information and Comments

I enjoyed addressing this bug and contributing to the ongoing refinement of the Manim library. This small enhancement will not only resolve an open issue but also make the library more robust and user-friendly.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Review by behackl — CHANGES_REQUESTED** [review]
There are a few things that should be improved here; currently this PR adds too much overhead.

My recommendation is to not introduce separate methods checking for empty text. If required, they should not live in the animation class, but rather in their dedicated text classes, i.e., `MarkupText.is_empty` and `Text.is_empty` -- but actually, this isn't really required here, we can just check whether the passed mobject has any points.

In fact, I'd suggest to simply add a check for the empty text before the `run_time` check, without any further changes to the implementation. You should then add a new (non-graphical) test to a new file `tests/module/animation/test_creation.py` (look at the other files in that directory to get an idea of how the file should look like) to make sure the error is raised correctly in the two cases you've described here.

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
Thank you for the information @behackl !
I have failed to noticed a lot of things that are wrong and thanks to your comments I am able to learn and hopefully not repeat them in future mistakes haha!

I will try to fix my comments Ben! Thank you for the help!

**Comment by behackl** [comment]
Sure, just ask in case anything is unclear.

In case you intend to continue working on this issue, you can just keep pushing changes to this pull request; no need to close this and open another.

**Review by Immanuel-Alvaro-Bhirawa — COMMENTED** [review]
Hi Ben,
I have made changed to the PR and to the manim code according to your comments!
I hope it's correct now. I have tested the code and it has run as the user wanted in the github issue.

If there are anything that is still lacking, I kindly request another comment if it doesn't take too much of your time.

Kind regards,
Immanuel Alvaro Bhirawa

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
Hi @behackl ,
I have updated both the manim repo code for `AddTextLetterByLetter` and the PR overview comments according to your instructions.

If there are any that is still lacking, I kindly request another comment from you if it doesn't take too much of your time!
Kind regards,
Immanuel Alvaro Bhirawa

**Review by Immanuel-Alvaro-Bhirawa — COMMENTED** [review]
Finished reviewing all the comments from Ben!

File: manim/animation/creation.py [hunk 1]
```diff
@@ -563,15 +563,22 @@ def __init__(
...
+
+        # Check for empty text using family_members_with_points()
+        if not text.family_members_with_points():
+            raise ValueError(
+                "The input text is empty i.e., input has no characters to read thus cannot render a video file."
+            )
+
         if run_time is None:
-            # minimum time per character is 1/frame_rate, otherwise
-            # the animation does not finish.
```
Comment (behackl) [inline manim/animation/creation.py:568]: any reason for removing this comment?

File: manim/animation/creation.py [hunk 2]
```diff
@@ -563,15 +563,22 @@ def __init__(
...
             run_time = np.max((1 / config.frame_rate, self.time_per_char)) * len(text)
+
```
Comment (behackl) [inline manim/animation/creation.py:575]: Any reason for adding this newline?

File: manim/animation/creation.py [hunk 3]
```diff
@@ -563,15 +563,22 @@ def __init__(
...
         super().__init__(
             text,
             suspend_mobject_updating=suspend_mobject_updating,
             int_func=int_func,
             rate_func=rate_func,
+            time_per_char=time_per_char,
```
Comment (behackl) [inline manim/animation/creation.py:581]: The parent class, `ShowIncreasingSubsets` does not seem to have a `time_per_char` keyword argument -- this should not be propagated upwards.

File: manim/animation/creation.py [hunk 4]
```diff
@@ -563,15 +563,22 @@ def __init__(
         **kwargs,
     ) -> None:
         self.time_per_char = time_per_char
+
+        # Check for empty text using family_members_with_points()
+        if not text.family_members_with_points():
+            raise ValueError(
+                "The input text is empty i.e., input has no characters to read thus cannot render a video file."
```
Comment (behackl) [inline manim/animation/creation.py:570]: Personally I'd prefer a shorter error message (e.g., `f"The text mobject {text} does not seem to contain any characters."`) and remove the comment about the video file (which is a consequence the user does not strictly need to know about).

**Review by behackl — CHANGES_REQUESTED** [review]
Thanks for the update! I've left a few more comments -- and the test that I've mentioned in my last review would still be nice to have.

File: manim/animation/creation.py (hunk 1 above)
Comment (Immanuel-Alvaro-Bhirawa) [reply manim/animation/creation.py:568]: Oh that's is my bad Ben!

File: manim/animation/creation.py (hunk 2 above)
Comment (Immanuel-Alvaro-Bhirawa) [reply manim/animation/creation.py:575]: No reason at all!
I wasn't paying attention again. Completely My bad.

File: manim/animation/creation.py (hunk 3 above)
Comment (Immanuel-Alvaro-Bhirawa) [reply manim/animation/creation.py:581]: Understood got it!

File: manim/animation/creation.py (hunk 4 above)
Comment (Immanuel-Alvaro-Bhirawa) [reply manim/animation/creation.py:570]: Okay got it Ben!

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
Hi @behackl , I have a question in regards to the test file that you requested.
It appears in my local machine, I can't test the new test file `tests/module/animation/test_creation.py` as it appears to say `pytest` is not recognised / installed even though it's already imported.

Thus, should I send the test codes here for you to check them instead?
Should the answer be yes, here are the codes for `test_creation.py`
```from __future__ import annotations

import pytest
from manim import Text, AddTextLetterByLetter, config
import numpy as np


def test_non_empty_text_creation():
    """Check if AddTextLetterByLetter works for non-empty text."""
    s = Text("Hello")
    anim = AddTextLetterByLetter(s)
    assert anim.mobject.text == "Hello"


def test_empty_text_creation():
    """Ensure ValueError is raised for empty text."""
    with pytest.raises(ValueError, match="does not seem to contain any characters"):
        AddTextLetterByLetter(Text(""))


def test_whitespace_text_creation():
    """Ensure ValueError is raised for whitespace-only text, assuming the whitespace characters have no points."""
    with pytest.raises(ValueError, match="does not seem to contain any characters"):
        AddTextLetterByLetter(Text("    "))


def test_run_time_for_non_empty_text():
    """Ensure the run_time is calculated correctly for non-empty text."""
    s = Text("Hello")
    run_time_per_char = 0.1
    expected_run_time = np.max((1 / config.frame_rate, run_time_per_char)) * len(s.text)
    anim = AddTextLetterByLetter(s, time_per_char=run_time_per_char)
    assert anim.run_time == expected_run_time```

**Comment by behackl** [comment]
> it appears to say `pytest` is not recognised / installed

Did you setup your development environment as described in out documentation? As long as you are in your poetry shell, pytest should be available to you.

I do not want to test your changes beforehand, no -- just push the file to this branch, the CI pipeline will carry out the tests automatically if you can't test them locally. Thank you!

**Comment by behackl** [comment]
I've also just unresolved the previous conversations as long as the changes have not been carried out yet.

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
> > it appears to say `pytest` is not recognised / installed
> 
> Did you setup your development environment as described in out documentation? As long as you are in your poetry shell, pytest should be available to you.
> 
> I do not want to test your changes beforehand, no -- just push the file to this branch, the CI pipeline will carry out the tests automatically if you can't test them locally. Thank you!

Okay Ben Thank you!
Got it!!

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
> Did you setup your development environment as described in out documentation? As long as you are in your poetry shell, pytest should be available to you

Ah I didnt.. I was running it on the VScode terminal.
Understood, I got it Ben!

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
> I've also just unresolved the previous conversations as long as the changes have not been carried out yet.

Got It! I'm sorry for being too early on resolving conversations.

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
Hi @behackl ! 
I have made the requested changes, from removing comments and changing exception string message to making a new test file in `tests/module/animation/test_creation.py`.

I kindly request another review, should it be possible. I would like to apologize as well for keep requesting for your reviews and disturbing your time.

If everything's good and adjusted, I will mark the conversations above as solved. And, if possible, the changes that I made are merged to the main branch.

Again, thank you Ben!

**Review by behackl — APPROVED** [review]
This looks fine now! Thank you for your contribution!

**Comment by Immanuel-Alvaro-Bhirawa** [comment]
Thank you so much Ben for all the help!!!

**Pull request merged**